### PR TITLE
REF: call `obj.describe()` in a thread pool; cache descriptions + widget types

### LINKED ISSUE #273: Revisit original SignalPanel destroy-and-recreate functionality
The SignalPanel creates only signals requested by the user.

If the user-requested signal kinds are changed, the widgets are destroyed and new ones are re-created.

The new CompositeSignalPanel selects _everything_ and then shows/hides based on user filters.

(Path forward not clear just yet - this is more of a reminder than anything)

### PR DESCRIPTION
**Pull request by klauer**
Overview
----------
Calls `describe` in the Qt thread pool

Details
-------
* `describe()` calls happen in a separate thread to determine widget classes
* A global `describe` cache can be used to not require multiple calls to `signal.describe()` (between different signal panels)
* Additionally, a global cache holds signal->widget type information based on the describe cache
* Rows are reserved for all components, regardless of filter settings
   - A to-be-shown signal will be instantiated and displayed
   - Otherwise, the row label is hidden and awaits a change of filters

Closes #273 with the component refactor

TODO
------
- [x] Further verification it does what we want
- [x] Failing tests
- [x] Targeted tests for the descriptor cache
- [x] Docstrings
- [x] `create_signal_widget` unused - remove or deprecate?
- [x] `filter_settings` pulled from widget -> layout indicate that methods should be moved around

### CODE REVIEW

**Comment by klauer** [comment]
(As I'm making another mess here, if I ever get this out of draft form, I'll just ask the reviewers to give this a try and not review in-depth.)

**Comment by klauer** [comment]
@ZLLentz if you could give this a try with a couple screens as you did in the previous PR, it would be appreciated. If bored/ambitious, the code is there as well for bonus points...

**Comment by ZLLentz** [comment]
@klauer I'll give this a try now, with the crazy screen and with a few more normal screens

**Comment by ZLLentz** [comment]
This has a noticeable speedup on the not-fake PPM screen (40s vs 60s to load). Interestingly this also makes the real PPM load faster than the fake PPM.  (The PPM class is slated to be decreased in size so hopefully that will help with the remaining time chunk)

I ran into various bugs involving clicking the buttons at the top that switch between embedded/tree/engineering templates, these showed exceptions in the terminal and made the screen go blank:

Using `typhos --fake-device "pcdsdevices.device_types.PPM[]"` and trying to go to embedded:
<details>
Traceback (most recent call last):
  File "/u/lu/zlentz/pydev/typhos/display.py", line 356, in _template_selected
    self.device_display.force_template = template
  File "/u/lu/zlentz/pydev/typhos/display.py", line 838, in force_template
    self.load_best_template()
  File "/u/lu/zlentz/pydev/typhos/display.py", line 769, in load_best_template
    self._update_children()
  File "/u/lu/zlentz/pydev/typhos/display.py", line 824, in _update_children
    widget.add_device(device)
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 566, in add_device
    self._update_panel()
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 513, in _update_panel
    self._panel_layout.filter_signals(**self.filter_settings)
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 375, in filter_signals
    self._set_visible(name, visible)
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 350, in _set_visible
    info['signal'] = signal = create_func()
  File "/u1/zlentz/miniconda3/envs/dev/lib/python3.6/site-packages/ophyd/device.py", line 1197, in __getattr__
    raise AttributeError(name)
AttributeError: enable
</details>

Using `typhos --fake-device "pcdsdevices.device_types.BeckhoffAxis[]"` and trying to go to embedded:
<details>
Traceback (most recent call last):
  File "/u/lu/zlentz/pydev/typhos/display.py", line 136, in selected
    setattr(panel, prop, new_value)
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 501, in _set_kind
    self._update_panel()
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 513, in _update_panel
    self._panel_layout.filter_signals(**self.filter_settings)
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 375, in filter_signals
    self._set_visible(name, visible)
  File "/u/lu/zlentz/pydev/typhos/panel.py", line 350, in _set_visible
    info['signal'] = signal = create_func()
  File "/u1/zlentz/miniconda3/envs/dev/lib/python3.6/site-packages/ophyd/device.py", line 1197, in __getattr__
    raise AttributeError(name)
AttributeError: cmd_err_reset
</details>

The real PPM worked for a while, but eventually failed in the same way after enough persistent clicking. Sometimes when it did work a bunch of labels went missing in embedded mode.

I also had the following startup error for a fake `BeckhoffAxis`:
`[2020-05-01 13:17:52] - ERROR -  Unable to find signal for <PyDMChannel (sig://device)> in signal registry.Use typhos.plugins.register_signal()`

File: tests/test_signal.py [hunk 1]
```diff
@@ -1,24 +1,72 @@
 import random
```
Comment (ZLLentz) [inline tests/test_signal.py:1]: Is this file due for a rename to `test_panel.py`?

Comment (klauer) [reply tests/test_signal.py:1]: Yes, I can include that here, too.

**Comment by ZLLentz** [comment]
I tried to review the code, but once again I'm not familiar enough with qt things and you've made a ton of changes... so I don't really have any input right now

**Comment by klauer** [comment]
Thanks for giving it a try. Will look into those issues.

I hope that this PR in combination with #287 will further decrease load times.

**Comment by hhslepicka** [comment]
Here is what I could find while trying to reproduce Zach's report:

Test Case 1
------------
The `typhos --fake-device "pcdsdevices.device_types.PPM[]"` took a minute for me locally to launch (but damn that is a big device).

When trying to go `embedded` I got the following:
```
Traceback (most recent call last):
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/display.py", line 356, in _template_selected
    self.device_display.force_template = template
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/display.py", line 838, in force_template
    self.load_best_template()
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/display.py", line 769, in load_best_template
    self._update_children()
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/display.py", line 824, in _update_children
    widget.add_device(device)
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/panel.py", line 571, in add_device
    self._update_panel()
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/panel.py", line 518, in _update_panel
    self._panel_layout.filter_signals(**self.filter_settings)
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/panel.py", line 380, in filter_signals
    self._set_visible(name, visible)
  File "/Users/slepicka/sandbox/git-pcdshub/typhon-git/typhos/panel.py", line 355, in _set_visible
    info['signal'] = signal = create_func()
  File "/Users/slepicka/sandbox/git-slepicka/ophyd-git/ophyd/device.py", line 1197, in __getattr__
    raise AttributeError(name)
AttributeError: enable
Abort trap: 6
```


Test Case 2
------------

The `typhos --fake-device "pcdsdevices.device_types.BeckhoffAxis[]"` took 1 second to launch. I also got the following error during launching:
```
[2020-05-01 14:36:47] - ERROR -  Unable to find signal for <PyDMChannel (sig://device)> in signal registry.Use typhos.plugins.register_signal()
```

Trying to go embedded worked perfectly.

**Comment by klauer** [comment]
2 fixes should address the attributeerror.
Not sure about the unregistered channel just yet.

**Comment by hhslepicka** [comment]
I can confirm that after the latest fix I can switch the monster PPM to embedded.

File: tests/test_signal.py (hunk 1 above)
Comment (klauer) [reply tests/test_signal.py]: Resolved in cf2f322

**Pull request merged**